lnwallet/musig_session.go

```diff
@@ -0,0 +1,552 @@
+// musig session.
+type musigSessionOpts struct {
+	// customRand is an optional custom random source that can be used to
+	// generate nonces via a counter scheme.
+	customRand io.Reader
+}
+
+// defaultMusigSessionOpts returns the default set of options for the musig
+// session.
+func defaultMusigSessionOpts() *musigSessionOpts {
+	return &musigSessionOpts{}
+}
+
+// MusigSessionOpt is a functional option that can be used to modify calls to
+// the musig session.
+type MusigSessionOpt func(*musigSessionOpts)
+
+// WithLocalCounterNonce is used to generate local nonces based on the shachain
+// producer and the current height. This allows us to not have to write secret
+// nonce state to disk. Instead, we can use this to derive the nonce we need to
+// sign and broadcast our own commitment transaction.
+func WithLocalCounterNonce(targetHeight uint64,
+	shaGen shachain.Producer) MusigSessionOpt {
+
+	return func(opt *musigSessionOpts) {
+		nextPreimage, _ := shaGen.AtIndex(targetHeight)
+
+		opt.customRand = bytes.NewBuffer(nextPreimage[:])
+	}
+}
+
+// VerifyCommitSig attempts to verify the passed partial signature against the
+// passed commitment transaction. A keyspend sighash is assumed to generate the
+// signed message. As we never re-use nonces, a new verification nonce (our
+// relative local nonce) returned to transmit to the remote party, which allows
+// them to generate another signature.
+func (m *MusigSession) VerifyCommitSig(commitTx *wire.MsgTx,
+	sig *lnwire.PartialSigWithNonce,
+	musigOpts ...MusigSessionOpt) (*musig2.Nonces, error) {
+
+	opts := defaultMusigSessionOpts()
+	for _, optFunc := range musigOpts {
+		optFunc(opts)
+	}
+
+	// Before we can verify the signature, we'll need to finalize the
+	// session by binding the remote party's provided signing nonce.
+	if err := m.FinalizeSession(musig2.Nonces{
+		PubNonce: sig.Nonce,
+	}); err != nil {
+		return nil, err
+	}
+
+	// When we verify a commitment signature, we always assume that we're
+	// verifying a signature on our local commitment. Therefore, we'll use:
+	// their remote nonce, and also public key.
+	partialSig := NewMusigPartialSig(
+		&musig2.PartialSignature{S: &sig.Sig},
+		m.nonces.SigningNonce.PubNonce, m.combinedNonce, m.signerKeys,
+	)
+
+	walletLog.Infof("verify partial sig: %v", spew.Sdump(partialSig))
+
+	// With the partial sig loaded with the proper context, we'll now
+	// generate the sighash that the remote party should have signed.
+	sigHash, err := taprootKeyspendSighash(
+		commitTx, m.inputTxOut.PkScript, m.inputTxOut.Value,
+	)
+	if err != nil {
+		return nil, err
+	}
+
+	walletLog.Infof("Verifying new musig2 sig for session=%x, nonce=%s",
+		m.session.SessionID[:], m.nonces.String())
+
+	if !partialSig.Verify(sigHash, m.remoteKey.PubKey) {
+		return nil, fmt.Errorf("invalid partial commit sig")
+	}
+
+	nonceOpts := []musig2.NonceGenOption{
+		musig2.WithPublicKey(m.localKey.PubKey),
+	}
+	if opts.customRand != nil {
```

Comment:
Hmm, how would we go about that? The `customRand` func in this case will use the new shachain root to derive nonces, also we'll start to pass in the txhash being signed into the nonce gen to further bind it. 

We only set the custom rand for our local commitment, which means we only sign with this nonce when we go to broadcast. 

Here's the argument I posted elsewhere for why imo the current approach is sound:

  1. we only use that verification nonce to sign when we broadcast
  2. we only ever have a single transaction valid to broadcast 
  3. after we sign and broadcast, we have that saved tx to rebroadcast, this tx never changes and we can't modify it (since we have another sig)
  4. leaking the nonce needs us to sign with that key w/ two diff messages, which itself would mean we're broadcasting a revoked state

Reply:
What I meant was, we'd make sure the opts is always set for `m.commitType == localCommit`, so maybe we can return an error `if opts.customRand == nil && m.commitType == localCommit`.

Reply:
By design it's meant to be an optional value though, we maintain similar control flow, but then allow the caller to specify this. I think a better area to add additional protections would be the channel state machine itself (part 6 in this series) where this is used. 

This would also break the current test that exercises the "normal" random nonce paths.